test/solvers/functional/functional_test.py

```diff
@@ -0,0 +1,602 @@
+                            places=PLACES)
+
+    # Test proximal operator. This requiers scaling to be positive.
+    sigma = 1.0
+    with pytest.raises(ValueError):
+        (neg_scal * F).proximal(sigma)
+
+    step_len = np.random.rand()
+    assert all_almost_equal(((scal * F).proximal(step_len))(x),
+                            (F.proximal(step_len * scal))(x),
+                            places=PLACES)
+
+    assert all_almost_equal(((F * scal).proximal(step_len))(x),
+                            ((1.0 / scal) *
+                                (F.proximal(step_len * scal**2)))(x * scal),
+                            places=PLACES)
+
+    # Test left multiplication with zero
+    assert isinstance(0 * F, odl.solvers.ZeroFunctional)
+
+
+def test_functional_composition():
+    """Test composition of functional.
+
+    This test tests composition, both from the right and from the left, with an
+    operator, which gives a functional, and with a vector, which returns an
+    operator."""
+
+    space = odl.uniform_discr(0, 1, 10)
+
+    func = odl.solvers.L2Norm(space)
+
+    # Test composition with operator from the right
+    scalar = np.random.rand()
+    wrong_space = odl.uniform_discr(1, 2, 10)
+    op_from_right_wrong = odl.operator.ScalingOperator(wrong_space, scalar)
+
+    with pytest.raises(TypeError):
+        func * op_from_right_wrong
+
+    op_from_right = odl.operator.ScalingOperator(space, scalar)
+    composition = func * op_from_right
+    assert isinstance(composition, odl.solvers.Functional)
+
+    x = noise_element(space)
+    op_in_x = op_from_right(x)
+    expected_result = func(op_in_x)
+    assert almost_equal(composition(x), expected_result, places=PLACES)
+
+    # Test gradient and derivative with composition from the right
+    grad_x = (op_from_right.adjoint * func.gradient *
+              op_from_right)(x)
+    assert all_almost_equal((composition.gradient)(x), grad_x, places=PLACES)
+
+    p = noise_element(space)
+    expected_result = grad_x.inner(p)
+    assert all_almost_equal(composition.derivative(x)(p), expected_result,
+                            places=PLACES)
+
+
+def test_functional_sum():
+    """Test for the sum of two functionals."""
+    space = odl.uniform_discr(0, 1, 10)
+
+    func1 = odl.solvers.L2NormSquared(space)
+    func2 = odl.solvers.L2Norm(space)
+
+    # Test for sum where one is not a functional
+    op = odl.operator.IdentityOperator(space)
+    with pytest.raises(TypeError):
+        func1 + op
+
+    # Test for different domain of the functionals
+    wrong_space = odl.uniform_discr(1, 2, 10)
+    func_wrong_domain = odl.solvers.L1Norm(wrong_space)
+    with pytest.raises(TypeError):
+        func1 + func_wrong_domain
+
+    func_sum = func1 + func2
+    x = noise_element(space)
+
+    # Test evaluation of the functionals
+    expected_result = func1(x) + func2(x)
+    assert almost_equal(func_sum(x), expected_result, places=PLACES)
```

Comment:
This can be written out, that way the error message is more clear if it fails and you need less temporary variables

```
assert almost_equal((func1 + func2)(x), 
                    func1(x) + func2(x), 
                    places=PLACES)
```